### PR TITLE
test for nsslapd-ignore-time-skew param of dirsrv in replica installation

### PR DESCRIPTION
**Pull request by mrizwan93**
value for nsslapd-ignore-time-skew should remain off by default for  
dirsrv. However while installing replica, it shuold be on. This test
checks that while installing replica, value for nsslapd-ignore-time-skew
set to on and after set to off.

related ticket : https://pagure.io/freeipa/issue/7211

Signed-off-by: Mohammad Rizwan Yusuf <myusuf@redhat.com>

### CODE REVIEW

File: ipatests/test_integration/test_replica_promotion.py [hunk 1]
```diff
@@ -527,3 +527,58 @@ def test_replica_install_with_existing_entry(self):
...
+        result = client.run_command(args)
+        assert result.returncode == 0
+
+        replica_install_log = client.get_file_contents(log_file)
+        occ_list = re.findall(regex, replica_install_log, re.M)
+        assert 'on' in occ_list[0] and 'off' in occ_list[1]
+
+        # check if nsslapd-ignore-time-skew is set to off for disrv
+        arg = ['ldapsearch', '-D',
+               str(replica.config.dirman_dn),   # pylint: disable=no-member
```
Comment (stlaz) [inline ipatests/test_integration/test_replica_promotion.py]: Why the `str` all of a sudden?

File: ipatests/test_integration/test_replica_promotion.py [hunk 2]
```diff
@@ -527,3 +527,58 @@ def test_replica_install_with_existing_entry(self):
...
+               '-w', replica.config.dirman_password,
+               '-h', replica.hostname,
+               '-b', 'cn=config']
+        result = replica.run_command(arg)
+        assert "nsslapd-ignore-time-skew: off" in result.stdout_text
+
+        # check on client promoted to replica
+        arg = ['ldapsearch', '-D',
+               str(replica.config.dirman_dn),   # pylint: disable=no-member
```
Comment (stlaz) [inline ipatests/test_integration/test_replica_promotion.py]: Why the str all of a sudden?

File: ipatests/test_integration/test_replica_promotion.py [hunk 3]
```diff
@@ -527,3 +527,58 @@ def test_replica_install_with_existing_entry(self):
...
+        tasks.install_replica(master, replica)
+        log_file = '/var/log/ipareplica-install.log'
+        replica_install_log = replica.get_file_contents(log_file)
+        regex = '^.*nsslapd\-ignore\-time\-skew:\n(.*)$'
+        occ_list = re.findall(regex, replica_install_log, re.M)
+        assert 'on' in occ_list[0] and 'off' in occ_list[1]
+
+        # check on replica promotion
+        tasks.install_client(master, client)
+        args = ['ipa-replica-install', '-U',
```
Comment (stlaz) [inline ipatests/test_integration/test_replica_promotion.py]: Wouldn't it make sense to have a task called e.g. `promote_to_replica()` instead and use it whenever needed throughout this file?

File: ipatests/test_integration/test_replica_promotion.py [hunk 4]
```diff
@@ -527,3 +527,58 @@ def test_replica_install_with_existing_entry(self):
...
+    def test_nsslapd_ignore_time_skew(self):
+        master = self.master
+        tasks.install_master(master)
+        replica = self.replicas[0]
+        client = self.replicas[1]
+
+        tasks.install_replica(master, replica)
+        log_file = '/var/log/ipareplica-install.log'
+        replica_install_log = replica.get_file_contents(log_file)
+        regex = '^.*nsslapd\-ignore\-time\-skew:\n(.*)$'
```
Comment (stlaz) [inline ipatests/test_integration/test_replica_promotion.py]: `regex = r'^.*nsslapd-ignore-time-skew:\n(.*)$'` ?

File: ipatests/test_integration/test_replica_promotion.py [hunk 5]
```diff
@@ -527,3 +527,58 @@ def test_replica_install_with_existing_entry(self):
...
+        occ_list = re.findall(regex, replica_install_log, re.M)
```
Comment (stlaz) [inline ipatests/test_integration/test_replica_promotion.py]: What's `occ`?
Also, `assert len(occ_list) == 2`

File: ipatests/test_integration/test_replica_promotion.py [hunk 6]
```diff
@@ -527,3 +527,58 @@ def test_replica_install_with_existing_entry(self):
...
+        tasks.install_client(master, client)
+        args = ['ipa-replica-install', '-U',
+                '-p', master.config.dirman_password,
+                '-w', master.config.admin_password,
+                '--ip-address', client.ip]
+        result = client.run_command(args)
+        assert result.returncode == 0
+
+        replica_install_log = client.get_file_contents(log_file)
+        occ_list = re.findall(regex, replica_install_log, re.M)
```
Comment (stlaz) [inline ipatests/test_integration/test_replica_promotion.py]: `assert len(occ_list) == 2`

**Comment by stlaz** [comment]
```
************* Module ipatests.test_integration.test_replica_promotion
ipatests/test_integration/test_replica_promotion.py:552: [E0602(undefined-variable), TestNsslapdIgnoreTimeSkew.test_nsslapd_ignore_time_skew] Undefined variable 're')
ipatests/test_integration/test_replica_promotion.py:552: [E0602(undefined-variable), TestNsslapdIgnoreTimeSkew.test_nsslapd_ignore_time_skew] Undefined variable 're')
ipatests/test_integration/test_replica_promotion.py:565: [E0602(undefined-variable), TestNsslapdIgnoreTimeSkew.test_nsslapd_ignore_time_skew] Undefined variable 're')
ipatests/test_integration/test_replica_promotion.py:565: [E0602(undefined-variable), TestNsslapdIgnoreTimeSkew.test_nsslapd_ignore_time_skew] Undefined variable 're')
```

File: ipatests/test_integration/test_replica_promotion.py (hunk 1 above)
Comment (mrizwan93) [reply ipatests/test_integration/test_replica_promotion.py]: when ran without str:
[ipa.ipatests.pytest_plugins.integration.host.Host.replica.OpenSSHTransport] RUN ['ldapsearch', '-D', ipapython.dn.DN('cn=Directory Manager'), '-w', 'Secret123', '-h', 'replica.testrelm.test', '-b', 'cn=config']
[ipa.ipatests.pytest_plugins.integration.host.Host.replica.cmd54] RUN ['ldapsearch', '-D', ipapython.dn.DN('cn=Directory Manager'), '-w', 'Secret123', '-h', 'replica.testrelm.test', '-b', 'cn=config']
FAILED[ipa.ipatests.pytest_plugins.integration.host.Host.master.OpenSSHTransport] RUN ['kinit', 'admin']

File: ipatests/test_integration/test_replica_promotion.py (hunk 5 above)
Comment (mrizwan93) [reply ipatests/test_integration/test_replica_promotion.py]: occ = occurrence.

 List contains no of occurrence for the searched param but we are interested in first 2 only.

**Comment by stlaz** [comment]
@mrizwan93 I did try to install a replica myself and view the log. I see that what you're doing here is only checking whether certain steps are being performed, not the actual state.

Although I do appreciate your contribution, I am afraid that there's no way on testing this properly right now. This test could just as well be but a grep of the source code and as such, it would not make sense.

**Comment by mrizwan93** [comment]
Thanks for the review.
Since this bug was about to set this attribute 'on' during replica install which we can find only in replica-install.log and we verified the bug in the same way.

**Comment by stlaz** [comment]
I see. Unfortunately that's not good enough for it to become a test, and, like I said, I don't think there's a way to test this right now. Thank you for the effort, though!

Closing this PR.

**Pull request closed without merging**